// @astrojs/og/lib/entrypoint.mjs
import { transform, walkSync, ELEMENT_NODE, TEXT_NODE } from "ultrahtml";
const { ImageResponse } = await import("./image-response.mjs");

async function resolveAsset(href, url) {
  if (href.startsWith("http")) {
    return fetch(href).then(res => res.text());
  }
  return fetch(new URL(href, url).toString()).then(res => res.text());
}
function inlineAssets({ url }) {
  return async function (doc) {
    const styles = [];
    const promises = [];
    const actions = [];
    walkSync(doc, (node, parent) => {
      if (node.name === "link" && node.attributes.rel === "stylesheet") {
        // ensure order is maintained
        const i = styles.push("");
        promises.push(
          resolveAsset(node.attributes.href, url).then(
            (css) => {
              if (css) {
                styles[i] = css;
                actions.push(() => {
                  parent.children = parent.children.filter((c) => c !== node);
                });
              }
            }
          )
        );
      }
    });
    await Promise.all(promises);
    for (const action of actions) {
      action();
    }
    doc.children.unshift({
      type: ELEMENT_NODE,
      name: "style",
      attributes: {},
      children: [
        {
          type: TEXT_NODE,
          value: styles.join("\n"),
        },
      ],
    });
    return doc;
  };
}
async function preprocess(markup, { url }) {
  return transform(markup, [inlineAssets({ url })]);
}

export const get = async ({ request, params }) => {
  const { name = "" } = params;
  const { origin, searchParams } = new URL(request.url);
  const config = { debug: false, width: 1200, height: 630 };
  for (const key of ["debug", "width", "height"]) {
    if (searchParams.has(key)) {
      const value = searchParams.get(key);
      if (key === "debug") {
        config[key] = value === "true" || value === "1";
      } else {
        config[key] = Number.parseInt(value);
      }
    }
    searchParams.delete(key);
  }
  const url = new URL(`/og${name ? `/${name}` : ""}?${searchParams}`, origin);
  const html = await fetch(url.toString())
    .then((res) => res.text())
    .then((text) => preprocess(text, { url }));

  return new ImageResponse(html, config);
};
