// @astrojs/og/lib/index.mjs
import fs from 'node:fs';

export default function og() {
  return {
    name: "@astrojs/og",
    hooks: {
      "astro:config:setup"({ injectRoute, updateConfig }) {
        injectRoute({
          pattern: "/_og/[...name]",
          entryPoint: "@astrojs/og/entrypoint",
        });
        updateConfig({
          vite: {
            plugins: [
              {
                name: '@astrojs/og/assets',
                enforce: 'pre',
                load(id) {
                  if (!id.endsWith('?asset')) return;
                  id = id.replace(/\?asset$/, '');
                  const buf = fs.readFileSync(id);
                  return `export default Buffer.from(${JSON.stringify(Array.from(buf))})`;
                }
              }
            ],
            ssr: {
              noExternal: ['@astrojs/og/entrypoint', '@astrojs/og/components'],
              external: [
                '@astrojs/og',
                "yoga-wasm-web",
                "@resvg/resvg-wasm",
                "satori",
                "satori-html",
              ],
            },
          },
        });
      },
    },
  };
}
